### PR TITLE
SI-6726 Improving pattern matcher analysis performance

### PR DESCRIPTION
**Pull request by retronym**
A dash of hash consing and a pinch of loop fusion.

These were just follow-the-profiler changes; I suspect there are bigger wins to be had. Perhaps the wins here don't merit closing the ticket.

Perhaps we could think about how to efficiently handle the common case where each case performs a single a) type test against a leaf type (one not subclassed by any other types that are tested), or b) equality test against a constant that is also unique within the match.

I'll leave consideration of that the reviewer, @adriaanm.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala [hunk 1]
```diff
@@ -1897,17 +1897,24 @@ trait PatternMatching extends Transform with TypingTransformers with ast.TreeDSL
...
+    abstract case class Sym(val variable: Var, val const: Const) extends Prop {
+      private[this] val id = Sym.nextSymId
+
       override def toString = variable +"="+ const +"#"+ id
     }
-    private def nextSymId = {_symId += 1; _symId}; private var _symId = 0
-
+    class UniqueSym(variable: Var, const: Const) extends Sym(variable, const)
+    object Sym {
+      private var uniques: util.HashSet[Sym] = new util.HashSet("uniques", 512)
```
Comment (xuwei-k) [inline src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala]: why var?

**Comment by adriaanm** [comment]
would be good to get rid of the gratuitous varrrrr, matey; then: LGTM --> thanks!

**Comment by adriaanm** [comment]
regarding the future work: yes, I agree that would be useful, as it's a very common case (ha)
i'm still hoping the switch to SAT4j and a more efficient encoding that avoids CNF would make this unnecessary
i'm not against it, though, just not very high on my list

**Comment by retronym** [comment]
PLS REBUILD ALL

**Comment by retronym** [comment]
Just realised I targetted this at master, not 2.10.x. @adriaanm please redirect if you wish.

**Pull request closed without merging**